### PR TITLE
List to Recycle

### PR DESCRIPTION
**Pull request by devilmonk**
I was busy for a while so this PR was due: [(https://github.com/kiwix/kiwix-android/pull/1448)]. I have done the changes that were mentioned there

I think it can still be improved like @macgills mentioned that he didn't like the KiwixFilter, and also that we want to have a new file for the adapter. 
Anyway, I think it is mergeable for now, and we should create a new Issue for these issues.

BTW, I am not really able to get why a new file for the adapter? Is there a more technical reason other than that it looks neat?

### CODE REVIEW

**Comment by kelson42** [comment]
@geekRishabhjain Which ticket is fixed by this PR?

**Comment by geekRishabhjain** [comment]
[https://github.com/kiwix/kiwix-android/issues/547](url) This is the issue on which this PR is targeted.

**Comment by macgills** [comment]
@geekRishabhjain yes it is about neatness. Neatness is essential, the more a class does/has in it the more it violates the principle of Single Responsibility, the harder it is to understand, the worse the code is. It is also more likely to cause issues because the boundaries between classes are muddied, inner classes implicitly hold a reference to their outer class and this is a recipe for disaster in android.
In addition there should not be 2 adapters for 1 recyclerview. An Adapter just renders data, it should accept any valid data.

**Comment by geekRishabhjain** [comment]
@macgills I didn't get what you meant by two adapters for one recycler view. You mentioned it in the previous PR too. Which other adapter are you talking of?


For the new file for the adapter: I actually replaced it at the same place it previously was, and further, I have always kept adapters in the same file(shame on me, though). I think that will form a good first issue.

**Comment by macgills** [comment]
`AutoCompleteAdapter` & `DefaultAdapter` you have 2 implementations of a Recyclerview Adapter in this PR. Ideally there will be just 1 and incorporate the adapter delegates pattern I outline in my [comment on the issue](https://github.com/kiwix/kiwix-android/issues/547#issuecomment-505437606).
The View in any MV_ should be a dumb data render.
I will hopefully get to reviewing this later today

**Comment by geekRishabhjain** [comment]
@macgills Yup, I got it. We can use AutoCompleteAdapter and get rid of this DefaultAdapter.
I think it was like this before I started working on it. Yes, this can be done.

**Comment by kelson42** [comment]
@geekRishabhjain What is the status here?

**Comment by geekRishabhjain** [comment]
@kelson42 No one has reviewed the PR so far. I thought that we were going to merge it, and create new issues: combine autocomplete adapter and default adapter.

**Comment by kelson42** [comment]
@macgills ?

**Comment by macgills** [comment]
I interpreted this
>We can use AutoCompleteAdapter and get rid of this DefaultAdapter.
I think it was like this before I started working on it. Yes, this can be done.

As "I will do this in this PR" which is ideally what I want to happen.

**Comment by geekRishabhjain** [comment]
@macgills Oh! I'll do it.

**Comment by geekRishabhjain** [comment]
@macgills I don't know where should I ask. Do we have(or have plans of) translation of pages.

I mean something like what google does. It translates the complete page into another language. If we are rendering websites, won't it be good to work on? @kelson42 What do you think?

**Comment by kelson42** [comment]
@geekRishabhjain Could we please focus on this PR (I don't see the relation beetween your question and this PR)? Kiwix is translated on translatewiki.net. Please open a new ticket if you have question related to other things.

**Comment by geekRishabhjain** [comment]
@kelson42 I am sorry about that.
Yes, I was getting it done. Just it came to my mind. I was wondering if non-English speaking Indians could use it. Anyway. Sorry again.

**Comment by geekRishabhjain** [comment]
@macgills 

I am struggling at getting it done. This is what I have tried:

1) I thought that I will get rid of DefaultAdapter completely, and at all places use autocompleteadapter. But then the two have different Viewholders.
Also there are some global variables that defaultadapter is using, I can't use them if I tried autocompleteadapter(it is in other file)

2) Then I tried to keep them different and calling autocompleteadapter's methods from default adapter, but that also did not work either.

3) I thought of extending default from autocomplete. I didn't find it worth even trying.

Can you please help me out, How should I proceed?
Should I create a completely new adapter? And extend these two from it?

I tried asking this question here, if you feel like please you can answer it: 

https://stackoverflow.com/questions/58716686/merging-two-mostly-similar-adapters-for-recyclerview

**Comment by macgills** [comment]
See my comment [here](https://github.com/kiwix/kiwix-android/issues/547#issuecomment-505437606)
>  Just be aware of the AdapterDelegates pattern used in LibraryAdapter, you can read more here, that covers the basics and I have taken it a bit further.

The key idea here is that we are displaying 2 different kinds of data, a history of search results and search suggestions from kiwixlib.

```
  private val libraryAdapter: LibraryAdapter by lazy {
    LibraryAdapter(
      BookDelegate(bookUtils, ::onBookItemClick), DividerDelegate
    )
  }
```
When I create the adapter here I add 2 `delegates`, a delegate is reponsible for identifying if data belongs to the delegate and then creating a viewHolder for that data, the viewholder is responsible for binding data/clickListeners to the views.

So what we really want is observable data and that will get rid of these global variables, if you look at how libraryAdapter gets updated

```
zimManageViewModel.libraryItems.observe(this, Observer(::onLibraryItemsChange))

private fun onLibraryItemsChange(it: List<LibraryListItem>?) {
    libraryAdapter.items = it!!
   // rest ommitted
  }
```

these topics can be quite complex and will involve Rx. If you are willing I can guide you through it but otherwise this works for now and I can get to reviewing it

**Comment by devilmonk** [comment]
I have exams coming up next week, I wanted these to be reviewed now, and I wanted work on the remainder of the code after my exams.

**Comment by devilmonk** [comment]
> these topics can be quite complex and will involve Rx. If you are willing I can guide you through it but otherwise this works for now and I can get to reviewing it.

I want to work on this problem, but have exams the next week. We can create a new issue, and I will solve them under your supervision if no one else does it till my exams finish.

File: core/src/main/java/org/kiwix/kiwixmobile/core/entity/MetaLinkNetworkEntity.java [hunk 1]
```diff
@@ -52,9 +52,9 @@ public FileElement getFile() {
     public List<Url> urls;
     @Element
     private long size;
-    @ElementMap(entry = "hash", key = "type", attribute = true, inline = true)
+    @ElementMap(entry = "hash", key = "type", attribute = true, inline = true, required = false)
     public Map<String, String> hashes;
-    @Element
+    @Element(required = false)
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/entity/MetaLinkNetworkEntity.java]: merge develop in, these are not your changes so I shouldn't see them

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 2]
```diff
@@ -56,18 +58,16 @@
...
 
   private final int REQ_CODE_SPEECH_INPUT = 100;
   @Inject
   SearchPresenter searchPresenter;
-  private ListView listView;
-  private ArrayAdapter<String> currentAdapter;
-  private AutoCompleteAdapter autoAdapter;
-  private ArrayAdapter<String> defaultAdapter;
+  private RecyclerView recyclerView;
+  AutoCompleteAdapter autoAdapter;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: package visibility?

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 3]
```diff
@@ -285,24 +266,58 @@ private void resetAdapter() {
...
+    return new DefaultAdapter();
+  }
 
-        if (convertView == null) {
-          row = LayoutInflater.from(parent.getContext())
-            .inflate(android.R.layout.simple_list_item_1, null);
-        } else {
-          row = convertView;
-        }
+  public class DefaultAdapter extends RecyclerView.Adapter<DefaultAdapter.ViewHolder> {
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: extract to a separate file

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 4]
```diff
@@ -285,24 +266,58 @@ private void resetAdapter() {
...
+
+    private Context context;
+    private AdapterView.OnItemClickListener onItemClickListener;
+    List<String> searchList;
+
+    @NonNull @Override
+    public DefaultAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+      View view = (View) LayoutInflater.from(parent.getContext())
+        .inflate(android.R.layout.simple_list_item_1, null);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Do not pass null to an inflate call, this can break theming as the newly inflated view should inherit attributes from its parent read [this](https://www.bignerdranch.com/blog/understanding-androids-layoutinflater-inflate/)
tl:dr in recycler do `inflate(R.layout.thing, parent, false)`

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 5]
```diff
@@ -285,24 +266,58 @@ private void resetAdapter() {
...
+    @NonNull public String getItem(int position) {
+      return searchList.get(position);
+    }
+
+    public class ViewHolder extends RecyclerView.ViewHolder {
+      TextView txt;
+
+      public ViewHolder(@NonNull View itemView) {
+        super(itemView);
+        txt = itemView.findViewById(android.R.id.text1);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Butterknife

File: settings.gradle [hunk 6]
```diff
@@ -1,4 +1,4 @@
 include ':core'
-include ':app'
 include ':custom'
+include ':app'
```
Comment (macgills) [inline settings.gradle]: why was this file changed?

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java [hunk 7]
```diff
@@ -139,11 +150,11 @@ protected FilterResults performFiltering(CharSequence constraint) {
 
     @Override
     protected void publishResults(CharSequence contraint, FilterResults results) {
-      mData = (ArrayList<String>) results.values;
+      data = (ArrayList<String>) results.values;
       if (results.count > 0) {
         notifyDataSetChanged();
       } else {
-        notifyDataSetInvalidated();
+        //notifyDataSetInvalidated();
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java]: it either needs to be here or it doesn't, leaving commented out code and an entire execution branch doesn't serve much of a purpose

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java [hunk 8]
```diff
@@ -82,9 +85,17 @@ public String getItem(int index) {
...
+  @Override public int getItemCount() {
+    return data.size();
+  }
+
+  public class ViewHolder extends RecyclerView.ViewHolder {
+    TextView title;
+
+    public ViewHolder(@NonNull View itemView) {
+      super(itemView);
+      title = itemView.findViewById(android.R.id.text1);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java]: Butterknife

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java [hunk 9]
```diff
@@ -34,45 +37,45 @@
...
 
-  @Override
-  public View getView(int position, View convertView, ViewGroup parent) {
-    View row = super.getView(position, convertView, parent);
-
-    TextView tv = row.findViewById(android.R.id.text1);
-    tv.setText(Html.fromHtml(getItem(position)));
+  @NonNull @Override public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+    View view = (View) LayoutInflater.from(parent.getContext())
+      .inflate(android.R.layout.simple_list_item_1, null);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java]: use the correct inflate method

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java [hunk 10]
```diff
@@ -34,45 +37,45 @@
...
+    ViewHolder holder = new ViewHolder(view);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/AutoCompleteAdapter.java]: this variable is unnecessary, you can just `return new ViewHolder(view)`

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java [hunk 11]
```diff
@@ -285,24 +266,58 @@ private void resetAdapter() {
...
+    @Override public int getItemCount() {
+      return searchList.size();
+    }
+
+    @NonNull public String getItem(int position) {
+      return searchList.get(position);
+    }
+
+    public class ViewHolder extends RecyclerView.ViewHolder {
+      TextView txt;
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: `txt` is a bit weak as names go. `recentSearch` or something like that would be better

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java (hunk 3 above)
Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Should I move the view holder too? I think yes, the doubt then arises. How should I go with updating searchPresenter.

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: Callbacks probably. Higher order functions if you use kotlin.

Comment (devilmonk) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: I don't use Kotlin
I found another way: some thing like this:
`if(context instanceof searchActivity){
((SearchActivity)this).searchPresenter ...
}`

I hope this will work fine.

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchActivity.java]: No, grabbing an instance an casting it is pretty messy. You can define an interface and supply an implementation in the constructor.

Consider `View.onSetClickListener`

```
public void setOnClickListener(@Nullable OnClickListener l)
```

You would have used this previously I imagine. What you do here is provide some code to be later executed when a view is clicked

```
    public interface OnClickListener {
        /**
         * Called when a view has been clicked.
         *
         * @param v The view that was clicked.
         */
        void onClick(View v);
    }
```

So we have to supply an implementation of this interface, note _implementation_ because an interface cannot be instantiated.
We can do this with an anonymous class
```
 view.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        //code that gets executed on a click
      }
    });
```
but it is handier if we can use the syntax sugar of a lambda

```
  view.setOnClickListener(v -> {
    //code that gets executed on a click
    });
```

There is nothing special about this, just another object, so you can extract a variable (`ctrl+alt+v`) and now it looks like this
```
    final View.OnClickListener onClickListener = v -> {

    };
    view.setOnClickListener(onClickListener);
```

Objects can of course be passed around so to other objects, composing functionality

```
class ContrivedExample {
  private final OnClickListener listener;
  public ContrivedExample(OnClickListener listener){
    this.listener=listener;
  }

  public bindClickListener(View v){
   v.setOnClickListener(listener) 
  }
}
```

which is very much what you want to do in the default adapter.

I also heavily recommend learning Kotlin, Android is Kotlin First and so is Kiwix.
